documents/design/slope_detection.md

```diff
@@ -0,0 +1,35 @@
+# Slope Detection Design Document
+
+**Author: Matthew Hannay**
+
+## Motivation
+The old `lidar_layer` worked for detecting barrels, but caused issues when approaching the ramp.
+The ramp would be seen as an obstacle, which meant we could no go past it.
+
+With slope detection and `traversability_layer`, a costmap will be generated based on the slope at any given point.
+For barrels, the steep walls will be detected as untraversable.
+For the ramp, the edges will be too steep to traverse, but the center will be detected as traversable.
+
+## Requirements
+Given an input pointcloud:
+1. Filter out far away points and points behind the robot.
+2. Add the points to an elevation map.
+3. Find the gradient at every point in the elevation map.
+4. Generate a costmap based on that gradient.
+
+## Design
+
+`pointcloud_filter` will filter out any points that are far away or behind the robot.
+
+With the `elevation_mapping` library, we can make a grid map with elevation in each cell. A node is created that
+subscribes to the filtered pointcloud and publishes an elevation map.
```

Comment:
In addition to talking about what the `elevation_mapping` does, I think it would be nice to
1. link to the related papers
2. add a link to the library on github
3. talk about how it is doing elevation mapping (ie. give a brief overview of the paper)
4. talk about why we chose to use this, and maybe some alternatives that were also considered.